### ISSUE TITLE
Hashgraph Branch Detector

### ISSUE THREAD

**Issue by cody-littley** [hashgraph/hedera-services#13284]
Create a simple component that detects hashgraph branching.

# Output

For the time being, the output of the branch detector will be metrics and rate-limited logging. Eventually, the node may decide to take additional actions when it observes branching.

# Algorithm

- Create a map from `NodeId` to the event descriptor of the most recent event for each node.
   - the map contains the latest known non-ancient event descriptor from each node in the network
   - if there are no known non-ancient events from a creator, the map has a null entry  
- for each event in topological order
   - if the event's self parent is not the in the map, flag it as a branching event
   - make an exception if the self parent is ancient and the node has a null entry in the map

# Architecture

![branchDetector](https://github.com/hashgraph/hedera-services/assets/56973212/13e059eb-394a-4b05-b23f-fb3cabd7587d)

**Comment by poulok** [comment]
Would the metric just be a single counter for each branched event detected or a counter per node? In general I think we need to move away from per-node metrics. Perhaps a single counter is sufficient, paired with a log message that says which creator branched.

**Comment by cody-littley** [comment]
> Would the metric just be a single counter for each branched event detected or a counter per node? In general I think we need to move away from per-node metrics. Perhaps a single counter is sufficient, paired with a log message that says which creator branched.

We could do either. If a per-node metric is not desirable, we could just go with a single metric for all nodes plus a rate limited log.

**Comment by lpetrovic05** [comment]
What would the metric look like? number of nodes branching? if so, when would this value be reset?

**Comment by cody-littley** [comment]
@lpetrovic05 

> What would the metric look like? number of nodes branching? if so, when would this value be reset?

If it were just a single metric, then I'd just count the number of branches per second. For figuring out which nodes are branching I'd use logging. One rate limited logger per node configured to log at most once per 10 minutes. The metric would alert us to the branching, and we could use the logs to figure out how severe the branching was.

**Comment by lpetrovic05** [comment]
the rate limited logging might not be great. we would detect the first branch, log it, then silence for 10 min. we would not know how many nodes are branching

**Comment by cody-littley** [comment]
> the rate limited logging might not be great. we would detect the first branch, log it, then silence for 10 min. we would not know how many nodes are branching

I agree, could hide data. But it's probably the best we can do if we don't want to have one metric per node.

**Comment by cody-littley** [comment]
> Maybe we need to get more details about how we count a branch. Should we would increment a counter every time a new branch is created, and decrement it every time the tip of a branch goes ancient?

This significantly complicates the implementation. My suggestion is to do something simple so that we catch branching, and consider saving deeper anaysis for a future effort.

**Comment by rbair23** [comment]
Does this box need to get information from hashgraph as to which round is ancient?

**Comment by rbair23** [comment]
Modulo any details around how metrics/logs are counted and reported, the idea is a +1 from me. Will be handy now and the basis of what we need to do in the future.

**Comment by lpetrovic05** [comment]
I suggest a metric like `maxNonAncientWeight`. Keeps track of the maximum number of weight (as a fraction of total weight) at any given time, this can immediately tell us if an ISS was caused by branching

**Comment by cody-littley** [comment]
> Does this box need to get information from hashgraph as to which round is ancient?

Yes, we will wire it up to the `event window` wire. There is a bunch of stuff in the intake pipeline which currently subscribes to that data flow.

**Comment by cody-littley** [comment]
> I suggest a metric like maxNonAncientWeight. Keeps track of the maximum number of weight (as a fraction of total weight) at any given time, this can immediately tell us if an ISS was caused by branching

I will add this metric.
